src/_zkapauthorizer/resource.py

```diff
@@ -214,110 +213,174 @@ def _setup_replication(self, request):
         request.finish()
 
 
-@define
-class RecoverResource(Resource):
+class RecoverProtocol(WebSocketServerProtocol):
     """
-    Implement the endpoint for triggering local state recovery from a remote
-    replica.
+    Speaks the server side of the WebSocket /recover protocol.
+
+    A client connects to this to start recovery, sending an opening
+    message with the rquired capability.
+
+    As recovery is ongoing, the server sends status updates as they
+    become available.
+
+    When the recovery is finished, and final message is sent
+    (indicating overall success or failure) and the WebSocket is
+    closed.
     """
 
-    _log = Logger()
+    def onClose(self, wasClean, code, reason) -> None:
+        """
+        WebSocket API: we've lost our connection for some reason
+        """
+        try:
+            self.factory.clients.remove(self)
+        except ValueError:
+            # may not have initiated recovery yet so it might not be
+            # in the clients list
+            pass
+
+    def onMessage(self, payload, isBinary) -> None:
+        """
+        WebSocket API: a message has been received from the client (the
+        only thing they can send is a request to initiate recovery).
+        """
+        try:
+            body = loads(payload)
+            if set(body.keys()) != {"recovery-capability"}:
+                raise ValueError("Unknown keys present in request")
+            recovery_capability = from_string(body["recovery-capability"])
+            if not isinstance(recovery_capability, ReadonlyDirectoryURI):
+                raise ValueError("Not a readonly-dircap")
+        except Exception:
+            self.sendClose(
+                code=4000,
+                reason="Failed to parse recovery request",
```

Comment:
The previous implementation revealed more information about what the client has done wrong.  It would be nice to preserve that.